cub/device/dispatch/dispatch_reduce.cuh

```diff
@@ -55,898 +55,1291 @@ CUB_NAMESPACE_BEGIN
+                 bool debug_synchronous,
+                 int ptx_version)
+    : d_temp_storage(d_temp_storage)
+    , temp_storage_bytes(temp_storage_bytes)
+    , d_in(d_in)
+    , d_out(d_out)
+    , num_items(num_items)
+    , reduction_op(reduction_op)
+    , init(init)
+    , stream(stream)
+    , ptx_version(ptx_version)
+  {
+    CUB_DETAIL_RUNTIME_DEBUG_SYNC_USAGE_LOG
+  }
+
+  //---------------------------------------------------------------------------
+  // Small-problem (single tile) invocation
+  //---------------------------------------------------------------------------
+
+  /**
+   * @brief Invoke a single block block to reduce in-core
+   *
+   * @tparam ActivePolicyT
+   *   Umbrella policy active for the target device
+   *
+   * @tparam SingleTileKernelT
+   *   Function type of cub::DeviceReduceSingleTileKernel
+   *
+   * @param[in] single_tile_kernel
+   *   Kernel function pointer to parameterization of 
+   *   cub::DeviceReduceSingleTileKernel
+   */
+  template <typename ActivePolicyT, typename SingleTileKernelT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t
+  InvokeSingleTile(SingleTileKernelT single_tile_kernel)
+  {
+    cudaError error = cudaSuccess;
+    do
     {
-      CUB_DETAIL_RUNTIME_DEBUG_SYNC_USAGE_LOG
-    }
-
-
-    //------------------------------------------------------------------------------
-    // Small-problem (single tile) invocation
-    //------------------------------------------------------------------------------
-
-    /// Invoke a single block block to reduce in-core
-    template <
-        typename                ActivePolicyT,          ///< Umbrella policy active for the target device
-        typename                SingleTileKernelT>      ///< Function type of cub::DeviceReduceSingleTileKernel
-    CUB_RUNTIME_FUNCTION __forceinline__
-    cudaError_t InvokeSingleTile(
-        SingleTileKernelT       single_tile_kernel)     ///< [in] Kernel function pointer to parameterization of cub::DeviceReduceSingleTileKernel
+      // Return if the caller is simply requesting the size of the storage
+      // allocation
+      if (d_temp_storage == NULL)
+      {
+        temp_storage_bytes = 1;
+        break;
+      }
+
+      // Log single_reduce_sweep_kernel configuration
+      #ifdef CUB_DETAIL_DEBUG_ENABLE_LOG
+      _CubLog("Invoking DeviceReduceSingleTileKernel<<<1, %d, 0, %lld>>>(), "
+              "%d items per thread\n",
+              ActivePolicyT::SingleTilePolicy::BLOCK_THREADS,
+              (long long)stream,
+              ActivePolicyT::SingleTilePolicy::ITEMS_PER_THREAD);
+      #endif
+
+      // Invoke single_reduce_sweep_kernel
+      THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron(
+        1, ActivePolicyT::SingleTilePolicy::BLOCK_THREADS, 0, stream)
+        .doit(single_tile_kernel, d_in, d_out, num_items, reduction_op, init);
+
+      // Check for failure to launch
+      if (CubDebug(error = cudaPeekAtLastError()))
+      {
+        break;
```

Comment:
I am not a big fan of those break into return blocks. Could we simply do a `return error;`?

Reply:
I'd prefer consistency and small PRs in this case. We can have a cleanup PR later that would remove all `do{} while(0);` from the code.